README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# cecmidas

<!-- badges: start -->
<!-- badges: end -->

cecmidas is an R package designed for interaction and uploading to the
California Energy Commission’s (CEC) Market Informed Demand Automation
Server (MIDAS). MIDAS provides access to real-time and historical
electricity rates for the largest electricial utilities in California.
Documentation for MIDAS is available at:
[MIDAS-Documentation](https://gitlab.com/CEC-MIDAS/midas-documentation).
Reading the MIDAS documentation is highly recommended as it makes using
this package much easier.

cecmidas can both GET and POST to MIDAS, but most users will only have
GET access. POST access is limited to the California utilities that
maintain their rates in MIDAS. The package includes some helper
functions for those utilities to help prep and upload their rate data.

This package is under active development. One of the next priorities is
to add a way for handling secrets so that you don’t expose passwords in
your code. If you have issues or feature requests, please use [GitHub
issues](https://github.com/stefwayland/cecmidas/issues) to provide
feedback.

## Installation

<!-- You can install the released version of cecmidas from [CRAN](https://CRAN.R-project.org) with: -->
<!-- ``` r -->
<!-- install.packages("cecmidas") -->
<!-- ``` -->

You can install the development version from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("stefwayland/cecmidas")
```

## Getting Started

Using MIDAS requires that you first register with the server. You only
need to do this once. MIDAS will send a conifrmation email to the
address you give it. Click through the confirmation link and you will be
good to go! (Unless you are a utility in California who needs upload
access, then see the MIDAS documentation for next steps.)

cecmidas uses [R6](https://r6.r-lib.org/) objects, so the code here
looks a little different than standard R code. The first step to using
cecmidas is to create a `midas` object (you can name it whatever you
want) which holds your login information and connection token. This
makes it easier to use once you get started because you don’t have to
provide that information to every function or store it in an
environment. To do this yourself, just replace the username etc., in the
example with your own.

``` r
library(cecmidas)
# Create a MIDAS object
midas <- MIDAS$new(username = "alovelace",
                   password = "Differenceengineftw!",
                   email = "alovelace@example.com",
                   fullname = "Ada Lovelace",
                   organization = "First Computing Inc")
# Register
midas$register()
```

Once you’ve registration process the fist time, you will not have to use
the `register()` function again. Just create a midas object and get on
with it. For example, to get the list of the rates available and then
look up the current price for one of those rates:

``` r
library(cecmidas)
# Create a MIDAS object (only do this one per session)
midas <- MIDAS$new(username = "alovelace",
                   password = "Differenceengineftw!")

# Get list of Rate Identification Numbers (RINs) and put them in a data.frame
rins <- midas$rins()
rins

# Get real-time data for the rate with RIN "USCA-TSTS-TTOU-TEST"
prices <- midas$value(rin = "USCA-TSTS-TTOU-TEST", query_type = "realtime")
prices$ValueInformation
```


NEWS.md
# cecmidas 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(cecmidas)

test_check("cecmidas")


tests/TOU to streaming.R
# TOU to XML converter

# DOW
# Holidays
library(lubridate)
library(data.table)

holidays <- fread("./tests/Holiday_test.csv", colClasses = c("character", "IDate"))
tou <- fread("./tests/TOU_test.csv", colClasses = c(DateStart = "IDate", DateEnd = "IDate",
                                                    TimeStart = "ITime", TimeEnd = "ITime"))

# Create table of prices for all hours of the year for each rate -------------
prices <- CJ(RIN = unique(tou$RIN),
             DateStart = seq(as.IDate("2023-01-01"), as.IDate("2023-12-31"), by = "day"),
             TimeStart = seq(as.ITime("00:00"), as.ITime("23:00"), by = 3600))
# Set Day to equal Monday = 1 through Sunday = 7
prices[, Day := wday(DateStart) - 1]
prices[Day == 0, Day := 7]
# Set Day to 8 on holidays
prices[holidays, Day := 8, on = .(DateStart = date)]

p2 <- tou[prices, .(RIN, RateName, RateType, DateStart, TimeStart, Day, Value, Unit), on = .(RIN = RIN, DateStart <= DateStart, DateEnd >= DateStart, DayStart <= Day, DayEnd >= Day, TimeStart <= TimeStart, TimeEnd >= TimeStart)]


# Create datetime fields for conversion to UTC
p2[, starttime := as.POSIXct(DateStart, time = TimeStart, tz = "America/Los_Angeles")]
p2[, c("DateStart", "TimeStart") := IDateTime(with_tz(starttime, tz = "UTC"))]
p2[, c("DateEnd", "TimeEnd") := IDateTime(with_tz(starttime + 3599, tz = "UTC"))]


# Convert to XML for upload -------------------------------------------------
library(xml2)
# schema <- read_xml("./tests/MIDAS upload XSD.xsd")
RIN = "USCA-TSTS-HTOU-TEST"
rate_name = "CEC TEST24HTOU"
rate_type = "TOU"

rate = xml_new_root("DemandData")
xml_add_child(rate, "RateInformation")
xml_add_child(xml_child(rate), "RateID", RIN)
xml_add_child(xml_child(rate), "RateName", rate_name)
xml_add_child(xml_child(rate), "RateType", rate_type)
vi <- xml_add_child(xml_child(rate), "ValueInformation")

# xml_child(xml_child(rate), search = "ValueInformation")
for(i in 1:12) {
  vd <- xml_add_child(vi, "ValueData")

  xml_add_child(vd, "DateStart", as.character(p2[i, DateStart]))
  xml_add_child(vd, "DateEnd", as.character(p2[i, DateEnd]))
  xml_add_child(vd, "TimeStart", as.character(p2[i, TimeStart]))
  xml_add_child(vd, "TimeEnd", as.character(p2[i, TimeEnd]))
  xml_add_child(vd, "DayStart", p2[i, Day])
  xml_add_child(vd, "DayEnd", p2[i, Day])
  xml_add_child(vd, "ValueName", "price")
  xml_add_child(vd, "Value", p2[i, Value])
  xml_add_child(vd, "Unit", p2[i, Unit])
  print(i)
}

xml_validate(rate, schema)
write_xml(rate, file = "tests/Streaming_Test.xml", options =c("format"))



tests/testthat/test-midas.R
source("R/midas.R")
m <- MIDAS$new(username = "<username>",
               password = "<password>",
               email = "<email>",
               fullname = "<full name>",
               organization = "<organization>")
# m$register()
crul::set_verbose()
m$get_token(new_token = TRUE)
m$midas_url = "http://midasapi.energy.ca.gov"
m$value(rin = "USCA-TSTS-HTOU-0000", query_type = "realtime", verbose = TRUE)
m$lookups(table_name = "Location", verbose = TRUE)
m$holiday()

# Test classic TOU upload
m$upload_rate(filename = "./tests/MIDAS_Test_Rate_24TOU.xml", verbose = TRUE)

# Test streaming rate upload with two rates
m$upload_rate(filename = "./tests/XML_upload.xml", verbose = TRUE)

# Test holiday upload
m$upload_holiday(filename = "./tests/MIDAS_Test_Holidays.xml", verbose = TRUE)


R/TOU to streaming.R
# cecmidas R package
# Software that connects to the CEC MIDAS Rate API
# Copyright (C) 2023 Stefanie Wayland
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

#
#' TOU to XML converter
#'
#' @description
#' Function to import CSV files that contain TOU rate information and convert
#' the TOU structure to a streaming structure.
#'
#' @param holiday_file atomic character with the path to a CSV file with holiday dates and information
#' @param tou_file atomic character with the path to a CSV file with TOU rate information
#' @param start_date atomic character or Date for the beginning of the period the streaming rate should cover
#' @param end_date atomic character or Date for the end of the period the streaming rate should cover
#' @param time_zone atomic character the time zone where the rate is offered. This should be one of the time zones listed in OlsonNames()
#'
#' @export
TOU_to_streaming <- function(holiday_file, tou_file, start_date,
                             end_date, time_zone = "America/Los_Angeles") {
  requireNamespace("lubridate")
  require(data.table)

  # check function arguments -------------------------------------------------
  checkmate::assert_file_exists(holiday_file, extension = "csv")
  checkmate::assert_file_exists(tou_file, extension = "csv")
  tryCatch(start_date <- as.IDate(start_date), error = function(e) stop("start_date must be a date formatted like YYYY-MM-DD"))
  tryCatch(end_date <- as.IDate(end_date), error = function(e) stop("end_date must be a date formatted like YYYY-MM-DD"))
  checkmate::assert_date(start_date, any.missing = FALSE, len = 1)
  checkmate::assert_date(end_date, any.missing = FALSE, len = 1)
  checkmate::assert_choice(time_zone, OlsonNames())

  # read in holiday and tou data and check column names ----------------------
  holidays <- fread(holiday_file, colClasses = c("character", "IDate"))
  checkmate::assert_names(names(holidays), permutation.of = c("holiday_name", "date"))
  tou <- fread(tou_file, colClasses = c(DateStart = "IDate", DateEnd = "IDate",
                                        TimeStart = "ITime", TimeEnd = "ITime"))
  checkmate::assert_names(names(tou),
                          permutation.of = c("RIN", "AltRateName1", "AltRateName2", "SignupCloseDate", "RateName",
                                             "RatePlan_Url", "RateType", "Sector", "API_Url", "EndUse", "DateStart",
                                             "DateEnd", "DayStart", "DayEnd", "TimeStart", "TimeEnd", "ValueName",
                                             "Value", "Unit"))

  # Create table of prices for all hours of the year for each rate -----------
  ## TODO: Consider adding Unit to unique query in case of time-varying demand charges or inclusion of other units in TOU rates
  prices <- CJ(RIN = unique(tou$RIN),
               DateStart = seq(as.IDate(start_date), as.IDate(end_date), by = "day"),
               TimeStart = seq(as.ITime("00:00"), as.ITime("23:00"), by = 3600))
  # Set Day to equal Monday = 1 through Sunday = 7
  prices[, DayType := wday(DateStart) - 1]
  prices[DayType == 0, DayType := 7]
  # Set Day to 8 on holidays
  prices[holidays, DayType := 8, on = .(DateStart = date)]

  p2 <- tou[prices, .(RIN, AltRateName1, AltRateName2, SignupCloseDate, RateName,
                      RatePlan_Url, RateType, Sector, API_Url, DateStart,
                      TimeStart, DayType, ValueName, Value, Unit),
            on = .(RIN = RIN, DateStart <= DateStart, DateEnd >= DateStart,
                   DayStart <= DayType, DayEnd >= DayType, TimeStart <= TimeStart, TimeEnd >= TimeStart)]


  # Create datetime fields and convert to UTC
  p2[, starttime := as.POSIXct(DateStart, time = TimeStart, tz = "America/Los_Angeles")]
  p2[, c("DateStart", "TimeStart") := IDateTime(lubridate::with_tz(starttime, tz = "UTC"))]
  # Drop hours due to DST changes
  # p2 <- unique(p2, by = c("RIN", "Unit", "DateStart", "TimeStart"))
  p2[, c("DateEnd", "TimeEnd") := IDateTime(lubridate::with_tz(starttime + 3599, tz = "UTC"))]
  p2
}


# Convert to XML for upload -------------------------------------------------

xml_non_na <- function(value, name) {
  if(!is.na(value)) {
    paste0("<", name, ">", value, "</", name, ">")
  }
}

#' Convert streaming rate to XML for upload
#'
#' @description
#' Function to convert the data.table output of the TOU_to_streaming function
#' into XML for upload to MIDAS.
#'
#' @param DT data.table created by the TOU_to_streaming function
#' @param file_name atomic character with the path where you want to save the XML
#'
#' @export
rate_to_xml <- function(DT, file_name) {
  rateinfo <- unique(DT[, .(RIN, AltRateName1, AltRateName2, SignupCloseDate, RateName,
                            RatePlan_Url, RateType, Sector, API_Url)])
  DemandData <- vector(mode = "list", length = nrow(rateinfo))
  for(rt in 1:nrow(rateinfo)) {
    RateInformation <- paste0(
      xml_non_na(rateinfo[rt, RIN], "RateID"),
      xml_non_na(rateinfo[rt, AltRateName1], "AltRateName1"),
      xml_non_na(rateinfo[rt, AltRateName2], "AltRateName2"),
      xml_non_na(rateinfo[rt, SignupCloseDate], "SignupCloseDate"),
      xml_non_na(rateinfo[rt, RateName], "RateName"),
      xml_non_na(rateinfo[rt, RatePlan_Url], "RatePlan_Url"),
      xml_non_na(rateinfo[rt, RateType], "RateType"),
      xml_non_na(rateinfo[rt, Sector], "Sector"),
      xml_non_na(rateinfo[rt, API_Url], "API_Url")
    )
    ratevalue <- DT[RIN == rateinfo[rt, RIN]]
    ValueInformation <- vector(mode = "list", length = nrow(ratevalue))
    for (vd in 1:length(ValueInformation)) {
      ValueData <- paste("<ValueData>",
                         paste0("<DateStart>", ratevalue[vd, DateStart], "</DateStart>"),
                         paste0("<DateEnd>", ratevalue[vd, DateEnd], "</DateEnd>"),
                         paste0("<TimeStart>", ratevalue[vd, TimeStart], "</TimeStart>"),
                         paste0("<TimeEnd>", ratevalue[vd, TimeEnd], "</TimeEnd>"),
                         paste0("<DayStart>", ratevalue[vd, DayType], "</DayStart>"),
                         paste0("<DayEnd>", ratevalue[vd, DayType], "</DayEnd>"),
                         paste0("<ValueName>", ratevalue[vd, ValueName], "</ValueName>"),
                         paste0("<Value>", ratevalue[vd, Value], "</Value>"),
                         paste0("<Unit>", ratevalue[vd, Unit], "</Unit>"),
                         "</ValueData>",
                         sep = "\n"
      )
      ValueInformation[[vd]] <- ValueData
    }
    ValueInformation <- paste("<ValueInformation>", paste(ValueInformation, collapse = "\n"), "</ValueInformation>", sep = "\n")
    RateInformation <- paste("<RateInformation>", RateInformation, ValueInformation, "</RateInformation>", sep = "\n")
    DemandData[[rt]] <- RateInformation
  }
  DemandData <- paste("<?xml version='1.0' encoding='UTF-8'?>",
                      "<DemandData>", paste(DemandData, collapse = "\n"), "</DemandData>", "\n", sep = "\n")

  writeChar(DemandData, file_name, eos = NULL)
}



R/TOU_to_streaming.R
# cecmidas R package
# Software that connects to the CEC MIDAS Rate API
# Copyright (C) 2023 Stefanie Wayland
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

#
# TOU to XML converter
#' TOU to streaming rate converter
#'
#' @description
#' Function to import CSV files that contain TOU rate information and convert
#' the TOU structure to a streaming structure.
#'
#' @param holiday_file atomic character with the path to a CSV file with holiday dates and information
#' @param tou_file atomic character with the path to a CSV file with TOU rate information
#' @param start_date atomic character or Date for the beginning of the period the streaming rate should cover
#' @param end_date atomic character or Date for the end of the period the streaming rate should cover
#' @param time_zone atomic character the time zone where the rate is offered. This should be one of the time zones listed in OlsonNames()
#'
#' @importFrom lubridate with_tz
#' @import data.table
#' @export
TOU_to_streaming <- function(holiday_file, tou_file, start_date,
                             end_date, time_zone = "America/Los_Angeles") {
  if (!requireNamespace("lubridate", quietly = TRUE)) {
    stop(
      "Package \"lubridate\" must be installed to use this function.",
      call. = FALSE
    )
  }
  if (!requireNamespace("data.table", quietly = TRUE)) {
    stop(
      "Package \"data.table\" must be installed to use this function.",
      call. = FALSE
    )
  }

  # check function arguments -------------------------------------------------
  checkmate::assert_file_exists(holiday_file, extension = "csv")
  checkmate::assert_file_exists(tou_file, extension = "csv")
  tryCatch(start_date <- as.IDate(start_date), error = function(e) stop("start_date must be a date formatted like YYYY-MM-DD"))
  tryCatch(end_date <- as.IDate(end_date), error = function(e) stop("end_date must be a date formatted like YYYY-MM-DD"))
  checkmate::assert_date(start_date, any.missing = FALSE, len = 1)
  checkmate::assert_date(end_date, any.missing = FALSE, len = 1)
  checkmate::assert_choice(time_zone, OlsonNames())

  # read in holiday and tou data and check column names ----------------------
  holidays <- fread(holiday_file, colClasses = c("character", "IDate"))
  checkmate::assert_names(names(holidays), permutation.of = c("holiday_name", "date"))
  tou <- fread(tou_file, colClasses = c(DateStart = "IDate", DateEnd = "IDate",
                                        TimeStart = "ITime", TimeEnd = "ITime"))
  checkmate::assert_names(names(tou),
                          permutation.of = c("RIN", "AltRateName1", "AltRateName2", "SignupCloseDate", "RateName",
                                             "RatePlan_Url", "RateType", "Sector", "API_Url", "EndUse", "DateStart",
                                             "DateEnd", "DayStart", "DayEnd", "TimeStart", "TimeEnd", "ValueName",
                                             "Value", "Unit"))

  # Create table of prices for all hours of the year for each rate -----------
  ## TODO: Consider adding Unit to unique query in case of time-varying demand charges or inclusion of other units in TOU rates
  prices <- CJ(RIN = unique(tou$RIN),
               DateStart = seq(as.IDate(start_date), as.IDate(end_date), by = "day"),
               TimeStart = seq(as.ITime("00:00"), as.ITime("23:00"), by = 3600))
  # Set Day to equal Monday = 1 through Sunday = 7
  prices[, DayType := wday(DateStart) - 1]
  prices[DayType == 0, DayType := 7]
  # Set Day to 8 on holidays
  prices[holidays, DayType := 8, on = .(DateStart = date)]

  p2 <- tou[prices, .(RIN, AltRateName1, AltRateName2, SignupCloseDate, RateName,
                      RatePlan_Url, RateType, Sector, API_Url, DateStart,
                      TimeStart, DayType, ValueName, Value, Unit),
            on = .(RIN = RIN, DateStart <= DateStart, DateEnd >= DateStart,
                   DayStart <= DayType, DayEnd >= DayType, TimeStart <= TimeStart, TimeEnd >= TimeStart)]


  # Create datetime fields and convert to UTC
  p2[, starttime := as.POSIXct(DateStart, time = TimeStart, tz = "America/Los_Angeles")]
  p2[, c("DateStart", "TimeStart") := IDateTime(lubridate::with_tz(starttime, tz = "UTC"))]
  # Drop hours due to DST changes
  # p2 <- unique(p2, by = c("RIN", "Unit", "DateStart", "TimeStart"))
  p2[, c("DateEnd", "TimeEnd") := IDateTime(lubridate::with_tz(starttime + 3599, tz = "UTC"))]
  p2[, .(RIN, AltRateName1, AltRateName2, SignupCloseDate, RateName,
         RatePlan_Url, RateType, Sector, API_Url, DateStart,
         TimeStart, DayStart = DayType, DayEnd = DayType, ValueName, Value, Unit)]
}

# Convert to JSON for upload ------------------------------------------------

#' Convert streaming rate to JSON for upload
#'
#' @description
#' Function to convert the data.table output of the TOU_to_streaming function
#' into JSON for upload to MIDAS.
#'
#' @param DT data.table created by the TOU_to_streaming function
#' @param file_name atomic character with the path where you want to save the JSON file
#' @param prettify atomic logical TRUE to generate prettified JSON, FALSE by default
#'
#' @import data.table
#' @export
rate_to_json <- function(DT, file_name, prettify = FALSE) {
  if (!requireNamespace("data.table", quietly = TRUE)) {
    stop(
      "Package \"data.table\" must be installed to use this function.",
      call. = FALSE
    )
  }
  rateinfo <- unique(DT[, .(RIN, AltRateName1, AltRateName2, SignupCloseDate, RateName,
                            RatePlan_Url, RateType, Sector, API_Url)])

  # Build list to convert to JSON
  rtl <- vector("list", length = nrow(rateinfo))
  for (rt in seq_along(rtl)) {
    rtl[[rt]] <- c(as.list(rateinfo[rt]),
                   list(ValueInformation = DT[RIN == rateinfo[rt, RIN],
                                              .(ValueName, DateStart, DateEnd,
                                                DayStart, DayEnd, TimeStart,
                                                TimeEnd, Value, Unit)]))
  }
  DemandData <- jsonlite::toJSON(rtl, dataframe = "rows", auto_unbox = TRUE, pretty = prettify)
  writeChar(DemandData, file_name, eos = NULL)
}


# Convert to XML for upload -------------------------------------------------

xml_non_na <- function(value, name) {
  if (!is.na(value)) {
    paste0("<", name, ">", value, "</", name, ">")
  }
}

#' Convert streaming rate to XML for upload
#'
#' @description
#' Function to convert the data.table output of the TOU_to_streaming function
#' into XML for upload to MIDAS.
#'
#' @param DT data.table created by the TOU_to_streaming function
#' @param file_name atomic character with the path where you want to save the XML
#'
#' @import data.table
#' @export
rate_to_xml <- function(DT, file_name) {
  if (!requireNamespace("data.table", quietly = TRUE)) {
    stop(
      "Package \"data.table\" must be installed to use this function.",
      call. = FALSE
    )
  }
  rateinfo <- unique(DT[, .(RIN, AltRateName1, AltRateName2, SignupCloseDate, RateName,
                            RatePlan_Url, RateType, Sector, API_Url)])
  DemandData <- vector(mode = "list", length = nrow(rateinfo))
  for(rt in 1:nrow(rateinfo)) {
    RateInformation <- paste0(
      xml_non_na(rateinfo[rt, RIN], "RateID"),
      xml_non_na(rateinfo[rt, AltRateName1], "AltRateName1"),
      xml_non_na(rateinfo[rt, AltRateName2], "AltRateName2"),
      xml_non_na(rateinfo[rt, SignupCloseDate], "SignupCloseDate"),
      xml_non_na(rateinfo[rt, RateName], "RateName"),
      xml_non_na(rateinfo[rt, RatePlan_Url], "RatePlan_Url"),
      xml_non_na(rateinfo[rt, RateType], "RateType"),
      xml_non_na(rateinfo[rt, Sector], "Sector"),
      xml_non_na(rateinfo[rt, API_Url], "API_Url")
    )
    ratevalue <- DT[RIN == rateinfo[rt, RIN]]
    ValueInformation <- vector(mode = "list", length = nrow(ratevalue))
    for (vd in 1:length(ValueInformation)) {
      ValueData <- paste("<ValueData>",
                         paste0("<DateStart>", ratevalue[vd, DateStart], "</DateStart>"),
                         paste0("<DateEnd>", ratevalue[vd, DateEnd], "</DateEnd>"),
                         paste0("<TimeStart>", ratevalue[vd, TimeStart], "</TimeStart>"),
                         paste0("<TimeEnd>", ratevalue[vd, TimeEnd], "</TimeEnd>"),
                         paste0("<DayStart>", ratevalue[vd, DayStart], "</DayStart>"),
                         paste0("<DayEnd>", ratevalue[vd, DayEnd], "</DayEnd>"),
                         paste0("<ValueName>", ratevalue[vd, ValueName], "</ValueName>"),
                         paste0("<Value>", ratevalue[vd, Value], "</Value>"),
                         paste0("<Unit>", ratevalue[vd, Unit], "</Unit>"),
                         "</ValueData>",
                         sep = "\n"
      )
      ValueInformation[[vd]] <- ValueData
    }
    ValueInformation <- paste("<ValueInformation>", paste(ValueInformation, collapse = "\n"), "</ValueInformation>", sep = "\n")
    RateInformation <- paste("<RateInformation>", RateInformation, ValueInformation, "</RateInformation>", sep = "\n")
    DemandData[[rt]] <- RateInformation
  }
  DemandData <- paste("<?xml version='1.0' encoding='UTF-8'?>",
                      "<DemandData>", paste(DemandData, collapse = "\n"), "</DemandData>", "\n", sep = "\n")

  writeChar(DemandData, file_name, eos = NULL)
}

